### ISSUE TITLE
DNS64/NAT64 Support in AdGuardHome?

### ISSUE THREAD

**Issue by gtxaspec** [AdguardTeam/AdGuardHome#5117]
### Prerequisites

- [X] I have checked the [Wiki](https://github.com/AdguardTeam/AdGuardHome/wiki) and [Discussions](https://github.com/AdguardTeam/AdGuardHome/discussions) and found no answer

- [X] I have searched other issues and found no duplicates

- [X] I want to request a feature or enhancement and not ask a question

### Description

Hello,

Didn't see this in any documentation, or in any previously opened issues or discussions.  Tested with latest AGH Edge release.

Regarding DNS64 support in AGH, I have IPv6 only clients ```(ex. 2001:db8:beef:4000::1)```, whom can access IPv4 hosts via a NAT64 jool translator ```(2001:db8:beef:4::/96)```.  

Traditionally, We would use `bind9` to specify a translator in `named.conf.options`, such as ```(dns64 2001:db8:beef:4::/96 { clients { !translator; dns64-good-clients; }; };```

The IPv6 only clients would then be able to query the DNS server, and if presented with a DNS entry that was IPv4 ONLY, the response would be `2001:db8:beef:4:8.8.8.8` for example.

I would ideally wish to replace bind9 with AdGuardHome.  

The question now is, does AdGuardHome support DNS64/NAT64?  If it does, how would it be configured?  If it does not, this would be a great feature to add to the existing features.

Thanks!

**Comment by jdreskell** [comment]
Following.  I used the guide at https://www.jool.mx/en/dns64.html to do this setup.  I'd like to use AGH instead of bind to serve my network.

**Comment by gtxaspec** [comment]
Looks liek this issue has been visited on the other AdGuard products:
AdGuard IOS: https://github.com/AdguardTeam/AdguardForiOS/issues/796
AdGuard Android: https://github.com/AdguardTeam/AdguardForAndroid/issues/1884#issuecomment-396570435
AdGuard dnsproxy: https://github.com/AdguardTeam/dnsproxy/issues/40

@ameshkov would you be able to look into this for AGH too?  =D

**Comment by ameshkov** [comment]
DNS64 is [supported by dnsproxy](https://github.com/AdguardTeam/dnsproxy#dns64-server) already which AdGuard Home uses under the hood. So this task may be split in two parts: expose DNS64 prefix setting via `AdGuardHome.yaml` (can be done relatively quickly) and expose it via the UI (later).

**Comment by gtxaspec** [comment]
@ameshkov great to hear!  I look forward to a future release sometime =D

**Comment by jdreskell** [comment]
Wonderful News!  @ameshkov is there a release version this would be planned in?  Cheers!

**Comment by gtxaspec** [comment]
tested latest master, working well.  can't wait for this to be introduced into the UI =D

**Comment by rapdodge** [comment]
^ same
tested on `v0.108.0-a.421+d52f1d0e`, working great, waiting to be implemented via webui :D

```
Server:  alpine-agsmartdns
Address:  192.168.1.130

Non-authoritative answer:
Name:    github.com
Addresses:  64:ff9b::14cd:f3a6
          20.205.243.166
```

**Comment by EugeneOne1** [comment]
@gtxaspec, @jdreskell, @rapdodge, hello.  We've filled the separate issue (#5460) about exposing the setting to UI.  We'll close this issue for now, so please upvote and track the mentioned one.

**Comment by Srylax** [comment]
I can't get this to work when the domain returns a CNAME:

`dig AAAA ipv4.google.com`
```
; <<>> DiG 9.10.6 <<>> AAAA ipv4.google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 40352
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;ipv4.google.com.		IN	AAAA

;; ANSWER SECTION:
ipv4.google.com.	300	IN	CNAME	ipv4.l.google.com.

;; AUTHORITY SECTION:
l.google.com.		2	IN	SOA	ns1.google.com. dns-admin.google.com. 583327338 900 900 1800 60

;; Query time: 60 msec
;; SERVER: 2001:XXXX:XXXX:XXXX::2#53(2001:XXXX:XXXX:XXXX::2) 
;; WHEN: Sat Nov 18 15:29:19 CET 2023
;; MSG SIZE  rcvd: 115
```

here with cloudflares dns64 server:
`dig @2606:4700:4700::64 AAAA ipv4.google.com`
```
; <<>> DiG 9.10.6 <<>> @2606:4700:4700::64 AAAA ipv4.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 56134
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;ipv4.google.com.		IN	AAAA

;; ANSWER SECTION:
ipv4.google.com.	123	IN	CNAME	ipv4.l.google.com.
ipv4.l.google.com.	123	IN	AAAA	64:ff9b::acd9:a82e

;; Query time: 53 msec
;; SERVER: 2606:4700:4700::64#53(2606:4700:4700::64)
;; WHEN: Sat Nov 18 15:31:11 CET 2023
;; MSG SIZE  rcvd: 93
```

When I query `ipv4.l.google.com`with dig it successfully returns the dns64 mapping but when I query `ipv4.google.com` via chromes net-internals it gives this response:
```
Resolved IP addresses of "ipv4.google.com": ["142.250.203.110"].
No alternative endpoints.
```

adguard query log:
```
15:35:56
ipv4.google.com
Type: AAAA, Plain DNS
Response:
CNAME: ipv4.l.google.com. (ttl=248)

15:35:55
ipv4.google.com
Type: A, Plain DNS
Response: 
CNAME: ipv4.l.google.com. (ttl=248)
A: 142.250.203.110 (ttl=248)

15:35:55
ipv4.google.com
Type: HTTPS, Plain DNS
Response: CNAME: ipv4.l.google.com. (ttl=7)
```

**Comment by Srylax** [comment]
Ok so turns out UDM is rewriting when the ad blocking feature is on: https://help.ui.com/hc/en-us/articles/9794438523799
This fucks up dns64...
With it turned off it works perfectly. 
Thank you!

**Comment by zbugrkx** [comment]
> Ok so turns out UDM is rewriting when the ad blocking feature is on: https://help.ui.com/hc/en-us/articles/9794438523799 This fucks up dns64... With it turned off it works perfectly. Thank you!

Hey!

Sorry for bumping a very old thread but that is the only one I could find about this same issue I have.

i"m also using a UI setup but do not have ad blocking enabled.

I have my vlans etc all pointing to my AGH for dns, same for dhcp.

When I enable dns64 in AGH with this settings in the yaml:

use_dns64: true
dns64_prefixes:

64:ff9b::/96

A windows machine when doing a ping to an ipv4 address gets the ipv6 prefix formated one but pings fail, likely because there is no nat64 and it does not get resolved right.

How did you get the whole setup to work with unifi? 

Thanks!

**Comment by Srylax** [comment]
> A windows machine when doing a ping to an ipv4 address gets the ipv6 prefix formated one but pings fail, likely because there is nessun nat64 and it does not get resolved right.
So that means dns64 is working correctly? For me I did not work because Unifi tampers with the dns resolution if the adblock is on.

It seems DNS64 is working for you but you're missing nat64? I use a single laptop which is in its own subnet which is the only client that owns a ipv4 address (besides the unifi stuff smh). The laptop runs https://www.jool.mx/en/index.html
Then everything with the `64:ff9b::/96` prefix gets redirected to this laptop (you have to configure this as a static route)

**Comment by zbugrkx** [comment]
> > A windows machine when doing a ping to an ipv4 address gets the ipv6 prefix formated one but pings fail, likely because there is nessun nat64 and it does not get resolved right.
> So that means dns64 is working correctly? For me I did not work because Unifi tampers with the dns resolution if the adblock is on.
> 
> It seems DNS64 is working for you but you're missing nat64? I use a single laptop which is in its own subnet which is the only client that owns a ipv4 address (besides the unifi stuff smh). The laptop runs https://www.jool.mx/en/index.html
> Then everything with the `64:ff9b::/96` prefix gets redirected to this laptop (you have to configure this as a static route)

AhB just to make sure i get this right.

That laptop of yours is your nat64 gateway then right?

The more i read and learn about all this the more it think my issue is nat64 as all i had setup was the dns64 and nothing else. So i guess the UDM then had no idea what to do with these ipv6 prefix ip's.

That static address you have is setup where? In the UDM? To tell it to send any ipv6 traffic from that prefix to the nat64 gateway?

Thanks a lot !!

**Comment by Srylax** [comment]
Yeah it is a bit tricky to wrap ones head around the first time. I think this is not the place to discuss this in detail. (I hope not too many people are subscribed to this issue). If you need more help you can message me at alive.mask3901 [at] hideme.ch

Nonetheless:

The OVERsimplified process is a bit like this:

1. Client makes dns request for `example.com`

- `if` response contains AAAA entry -> good just use the ipv6
- `if` response only contains A entry -> rewrite to use nat64 prefix -> send to router -> router static forwards to nat64 PLAT -> PLAT (needs ipv4) -> ipv4 website

dns64 is only needed if your host does not have a native CLAT. MacOS and windows (poorly implemented) have builtin CLAT and do not require CLAT (nat64 is still needed). Your mobile phone also has CLAT built in.
This is also important for application that do not use dns e.g. discord (i think they changed it) sent the raw ipv4 to the clients and so dns64 would not work here.

Devices that do not have a CLAT, are dependent on dns64 for their ipv6 resolution as they will only see "ah this website has an ipv6 address" and connect to it over ipv6 even if it is "only" the nat64 prefix.

so yes to answer your question you need to setup a server which is your nat64 and this server needs an ipv4 address. Furthermore you need to configure an static route in your UDM to forward all traffic to this machine with the `64:ff9b::/96` prefix
